### PR TITLE
UI: Fix flaky time-related tests

### PR DESCRIPTION
**Pull request by hashishaw**
## ✨ Cleanup Date usage ✨
This has caused headaches during testing, because the value of `new Date()` constantly changes due to the nature of time. With these changes, we no longer call `new Date()` within the application, and instead call a utility `timestamp.now()`. Under the hood, timestamp.now just returns `new Date()`, but it can be easily stubbed within tests to make the concept of "now" within tests is consistent and much easier to test. 

Another change made as part of this PR is replacing all instances of `new Date()` within tests (when used to generate a UID for engine and secret paths) with `uuidv4` which was added as part of 1.13. 

NOT part of these changes is swapping out `Date.now` throughout the codebase. This change is large enough as is, and we can revisit swapping that out at a later time

**How to review this**
Obviously a lot of files were changed, but most of the changes are fairly simple swaps. I recommend reviewing the code commit by commit, to see similar changes grouped together and larger refactors separated into their own commit.  **EDIT** things got sort of messy at the end there, trying to sort through test failures. Most of the files have 2-3 lines changed, so hopefully it's not too bad 😬 

### TL;DR How to test components that use time
Let's say you have a component that receives some timestamp from an API, and then renders something different depending on if the timestamp is in the past, future, close, far away, etc. Within the component, we would use `timestamp.now()` for the current time for comparison rather than `new Date()`. 

Then, the test might look something like this:

```js
import sinon from 'sinon';
import timestamp from 'core/utils/timestamp';
// ...other imports

module('Integration | Component | example-component', function (hooks) {
  setupRenderingTest(hooks);
  hooks.before(function() {
    // before all the module tests, we stub timestamp.now to return a static date of our choosing. 
    sinon.stub(timestamp, 'now').callsFake(() => new Date('2018-04-03T14:15:30'));
  })

  hooks.beforeEach(function() {
    // since timestamp.now is being called after we stubbed it, we know that the date will always return 2018-04-03T14:15:30.
    this.now = timestamp.now();
  })

  hooks.after(function() {
    // very important to reset the stub after these tests run, or it will still be stubbed for tests run afterward
    timestamp.now.restore();
  })

  test('it shows the right thing when date is in the future', function(assert) {
    // Now we can use a static date for which to compare to "now"
    this.set('apiDate', new Date('2019-01-01T12:30:00');
    await render(hbs`<ExampleComponent @apiDate={{this.apiDate}} />`)
    assert.dom(this.element).hasText('The API date is next year');
  });
});
```
If you want the value of `now` to change depending on the test, you could stub within the test. Note that if you do it this way but call `timestamp.now` in the `beforeEach` hook, the value of `timestamp.now` will not be stubbed at that point, and will give you the actual current time.

### CODE REVIEW

File: ui/tests/acceptance/client-dashboard-test.js [hunk 1]
```diff
@@ -11,23 +12,21 @@ import ss from 'vault/tests/pages/components/search-select';
...
 // upgrade happened 1 month after license start
-const UPGRADE_DATE = addMonths(LICENSE_START, 1);
+const UPGRADE_DATE = addMonths(LICENSE_START, 1); // 2022-08-01
 
 module('Acceptance | client counts dashboard tab', function (hooks) {
   setupApplicationTest(hooks);
   setupMirage(hooks);
 
   hooks.before(function () {
+    sinon.stub(Date, 'now').returns(staticNow);
```
Comment (hashishaw) [inline ui/tests/acceptance/client-dashboard-test.js:29]: Must stub in `hooks.before` not `hooks.beforeEach` or you get an error from sinon about wrapping a method which is already wrapped

File: ui/tests/acceptance/client-dashboard-test.js [hunk 2]
```diff
@@ -158,21 +154,18 @@ module('Acceptance | client counts dashboard tab', function (hooks) {
...
-    // query three months ago
-    const customEndDate = subMonths(CURRENT_DATE, 3);
+    // query three months ago (Oct 2022)
     await click(SELECTORS.rangeDropdown);
     await click('[data-test-show-calendar]');
-    if (parseInt(find('[data-test-display-year]').innerText) !== customEndDate.getFullYear()) {
-      await click('[data-test-previous-year]');
-    }
-    await click(find(`[data-test-calendar-month=${ARRAY_OF_MONTHS[customEndDate.getMonth()]}]`));
+    await click('[data-test-previous-year]');
```
Comment (hellobontempo) [inline ui/tests/acceptance/client-dashboard-test.js:167]: so much nicer 🥲

File: ui/tests/acceptance/oidc-provider-test.js [hunk 3]
```diff
@@ -125,13 +125,14 @@ module('Acceptance | oidc provider', function (hooks) {
   setupApplicationTest(hooks);
 
   hooks.beforeEach(async function () {
+    this.timestamp = new Date().getTime();
```
Comment (hellobontempo) [inline ui/tests/acceptance/oidc-provider-test.js:128]: So it looks like places where we're using a date to create a unique name or id you opted to use `new Date().getTime()` consistently? 

I like this! Makes it easy to discern when the date is relevant or just being used to timestamp and/or name a test item

Comment (hellobontempo) [reply ui/tests/acceptance/oidc-provider-test.js:128]: It seems like this was already the existing pattern...but I appreciate it! 😂

File: ui/tests/acceptance/secrets/backend/ssh/role-test.js [hunk 4]
```diff
@@ -12,11 +12,12 @@ module('Acceptance | secrets/ssh', function (hooks) {
   setupApplicationTest(hooks);
 
   hooks.beforeEach(function () {
+    this.timestamp = new Date().getTime();
     return authPage.login();
   });
 
-  const mountAndNav = async () => {
-    const path = `ssh-${new Date().getTime()}`;
+  const mountAndNav = async (assert, uid) => {
```
Comment (hellobontempo) [inline ui/tests/acceptance/secrets/backend/ssh/role-test.js:19]: out of curiosity - why does `assert` need to be passed here? I see it's passed below, but I don't see it used in the method itself?

File: ui/tests/integration/components/calendar-widget-test.js [hunk 5]
```diff
@@ -1,22 +1,22 @@
...
+const CURRENT_DATE = new Date('2020-03-15T14:15:00');
 
 module('Integration | Component | calendar-widget', function (hooks) {
   setupRenderingTest(hooks);
 
-  const isDisplayingSameYear = (comparisonDate, calendarYear) => {
-    return comparisonDate.getFullYear() === parseInt(calendarYear);
-  };
+  hooks.before(function () {
+    sinon.stub(Date, 'now').returns(CURRENT_DATE);
```
Comment (hellobontempo) [inline ui/tests/integration/components/calendar-widget-test.js:16]: NICE 🤩

**Review by hellobontempo — COMMENTED** [review]
Nice work on this so far! Thank you for tackling this cleanup work! Looking forward to less flaky tests 😍

File: ui/tests/acceptance/oidc-provider-test.js (hunk 3 above)
Comment (hashishaw) [reply ui/tests/acceptance/oidc-provider-test.js:128]: Exactly! Even in places where we only used it once in the test, hopefully by establishing this pattern we can keep things consolidated. Another option is to replace all of those with `uuidv4()` since that's a dependency now. I think both are valid directions, and then `timestamp` is less likely to be misused

File: ui/tests/acceptance/secrets/backend/ssh/role-test.js (hunk 4 above)
Comment (hashishaw) [reply ui/tests/acceptance/secrets/backend/ssh/role-test.js:19]: Good catch, I think it was leftover from a time when we were asserting within this block. I can remove it

File: ui/tests/acceptance/client-dashboard-test.js [hunk 6]
```diff
@@ -341,17 +334,14 @@ module('Acceptance | client counts dashboard tab', function (hooks) {
...
-        `Warning Vault was upgraded to 1.10.1 on ${format(
-          UPGRADE_DATE,
-          'MMM d, yyyy'
-        )}. We added monthly breakdowns and mount level attribution starting in 1.10, so keep that in mind when looking at the data. Learn more here.`
+        `Warning Vault was upgraded to 1.10.1 on Aug 1, 2022. We added monthly breakdowns and mount level attribution starting in 1.10, so keep that in mind when looking at the data. Learn more here.`
       );
   });
 
   test('Shows empty if license start date is current month', async function (assert) {
     // TODO cmb update to reflect new behavior
```
Comment (hashishaw) [inline ui/tests/acceptance/client-dashboard-test.js:342]: @hellobontempo is this TODO still relevant? The tests seem to pass but I'm not sure if this meant to check more things?

File: ui/tests/integration/helpers/date-format-test.js [hunk 7]
```diff
@@ -46,19 +54,26 @@ module('Integration | Helper | date-format', function (hooks) {
...
   test('displays time zone if withTimeZone=true', async function (assert) {
     const timestampDate = '2022-12-06T11:29:15-08:00';
-    const zone = new Date().toLocaleTimeString(undefined, { timeZoneName: 'short' }).split(' ')[2];
+    this.set('withTimezone', false);
     this.set('timestampDate', timestampDate);
 
-    await render(hbs`{{date-format this.timestampDate 'MMM d yyyy, h:mm:ss aaa' withTimeZone=true}}`);
-    assert.dom(this.element).hasTextContaining(`${zone}`);
+    await render(
+      hbs`<span data-test-formatted>{{date-format this.timestampDate 'MMM d yyyy, h:mm:ss aaa' withTimeZone=this.withTimezone}}</span>`
```
Comment (hashishaw) [inline ui/tests/integration/helpers/date-format-test.js:70]: Might have just been due to running this test locally, but I found that it was expecting "CST" and giving "CDT" (in tests run on main as well). Since the time just switched, I figured testing it this way (making sure there's the expected number of characters) ensures that daylight savings doesn't ruin this test in the future. Other suggestions welcome

File: ui/tests/acceptance/cluster-test.js [hunk 8]
```diff
@@ -31,15 +31,16 @@ const authAccessor = async function (path) {
...
+const setupUser = async function (uid) {
+  const authMethodPath = `userpass-${uid}`;
   await authAccessor(authMethodPath);
 };
 
 module('Acceptance | cluster', function (hooks) {
   setupApplicationTest(hooks);
 
   hooks.beforeEach(async function () {
+    this.timestamp = new Date().getTime();
```
Comment (kiannaquach) [inline ui/tests/acceptance/cluster-test.js:43]: same here - can we use the `ui/lib/core/addon/utils/timestamp.js `util here or is this supposed to be a static date since it's a test?

Comment (hashishaw) [reply ui/tests/acceptance/cluster-test.js:43]: We could -- one caveat is that if the test stubs timestamp then we need to use this syntax to get a non-static date. Timestamp.now in this scenario would be the same as new date so here it would be `this.timestamp.now().getTime()` which doesn't feel any cleaner. 

Another thing we could do, since it's just supposed to be a unique value, is to use `uuidv4()` instead. I'm leaning toward this along with more specific naming within the test so that it's easier to track down within the vault logs which test is triggering the log. What do you think of that?

Comment (kiannaquach) [reply ui/tests/acceptance/cluster-test.js:43]: I like the idea of using a `uuidv4()`.  It's also consistent with the pattern that's set up, making it easier to debug. :)

File: ui/tests/acceptance/client-dashboard-test.js (hunk 6 above)
Comment (hellobontempo) [reply ui/tests/acceptance/client-dashboard-test.js:342]: Oish yeah. I noticed this on my last passthrough and wondered the same thing 😅 

Among the miscellany of other bugs, I don't think I ever got an answer from the backend on what the expected response in this case (or perhaps forgot to ask). Previously data wasn't available for a customer until the first completed month, but that should no longer be the case.

This test can be deleted! But a good reminder of something that should be taken into account for the latest backend work on activity log testing

File: ui/tests/integration/helpers/date-format-test.js (hunk 7 above)
Comment (hellobontempo) [reply ui/tests/integration/helpers/date-format-test.js:70]: I think this makes sense!

**Comment by hashishaw** [comment]
> 



> So great to see these changes. Thank you for going after this. Quick question, do you think some kind of comment like you have in the PR description can be added to the pattern doc, with a link to this PR?

Absolutely! Great idea, thanks for the review
